### ISSUE TITLE
thread 'bulk_index' panicked at 'index 2497364531 out of range for slice of length 133729722', src/libcore/slice/mod.rs:2674:5

### ISSUE THREAD

**Issue by jlatone** [getumbrel/docker-electrs#5]
I'm running umbrel on an IBM Z (mainframe, 64-bit, big endian) running Ubuntu 20.04.2 LTS. Everything is running fine, except electrs. electrs effectively stops running (sleeps indefinitely) after the following error message:

> thread 'bulk_index' panicked at 'index 2497364531 out of range for slice of length 133729722', src/libcore/slice/mod.rs:2674:5

Starting from a new node, electrs runs and waits for bitcoin to fully sync:

> ...
> electrs | WARN - wait until IBD is over: headers=688177 blocks=688052 progress=0.9994805920510615
> electrs | WARN - wait until IBD is over: headers=688177 blocks=688058 progress=0.9995014055113616
> ...

once bitcoin is sync'd, electrs starts building the index:

> electrs | DEBUG - opening DB at "/data/db/mainnet"
> electrs | TRACE - latest indexed blockhash: 0000000000000000000000000000000000000000000000000000000000000000
> electrs | INFO - listing block files at "/data/.bitcoin/blocks/blk*.dat"
> electrs | INFO - indexing 2627 blk*.dat files
> electrs | DEBUG - found 0 indexed blocks
> electrs | TRACE - downloading 100000 block headers
> ...
> electrs | DEBUG - applying 688073 new headers from height 0
> electrs | TRACE - indexed "/data/.bitcoin/blocks/blk00000.dat": 0 rows
> electrs | TRACE - indexed "/data/.bitcoin/blocks/blk00001.dat": 0 rows
> electrs | TRACE - indexed "/data/.bitcoin/blocks/blk00002.dat": 0 rows

the following errors get logged and electrs sleeps forever after the last one:

> electrs | TRACE - indexed "/data/.bitcoin/blocks/blk00040.dat": 0 rows
> electrs | thread 'bulk_index' panicked at 'index 1457594409 out of range for slice of length 134058325', src/libcore/slice/mod.rs:2674:5
> ...
> electrs | TRACE - indexed "/data/.bitcoin/blocks/blk00077.dat": 0 rows
> electrs | thread 'bulk_index' panicked at 'index 3347378801 out of range for slice of length 133998011', src/libcore/slice/mod.rs:2674:5
> ...
> electrs | TRACE - indexed "/data/.bitcoin/blocks/blk00260.dat": 0 rows
> electrs | thread 'bulk_index' panicked at 'index 613586893 out of range for slice of length 133644501', src/libcore/slice/mod.rs:2674:5
> ...electrs | TRACE - indexed "/data/.bitcoin/blocks/blk00352.dat": 0 rows
> electrs | thread 'bulk_index' panicked at 'index 2497364531 out of range for slice of length 133729722', src/libcore/slice/mod.rs:2674:5

Any ideas what might be wrong, where I can start looking for the problem? I reproduced this several times, even trying different 3 different OS's to build the electrs image.

If it proves too elusive, can run something other than electrs?

Thanks for any pointers.

--joe

**Comment by lukechilds** [comment]
Would be worth rebuilding the container with the latest release of electrs before contacting them, I've just seen v0.8.10 is released and we're using v0.8.9 in Umbrel.

If that doesn't resolve the issue I'd suggest reporting this over on https://github.com/romanz/electrs since it looks like an issue specific to electrs and not Umbrel.

>If it proves too elusive, can run something other than electrs?

Yes! You can run ElectrumX. It takes quite a bit longer to sync than electrs but once it's done it's more performant.

**Comment by jlatone** [comment]
Thanks, @lukechilds. 

I just compiled & ran the latest electrs stand-alone on my native system and reproduced the problem. My guess is it's an endian issue, either with Rust or electrs. Looking at the source of the exception, with the code deserializing 32bit values from the bitcoin blocks, it may be a good guess. I closed the issue here and I'll spend a bit of time debugging it, though I may switch to ElectrumX regardless.

Did you happen to already dockerize ElectrumX to work with umbrel?

```
fn parse_blocks(blob: Vec<u8>, magic: u32) -> Result<Vec<Block>> {
    let mut cursor = Cursor::new(&blob);
    let mut blocks = vec![];
    let max_pos = blob.len() as u64;
    while cursor.position() < max_pos {
        let offset = cursor.position();
        match u32::consensus_decode(&mut cursor) {
            Ok(value) => {
                if magic != value {
                    cursor.set_position(offset + 1);
                    continue;
                }
            }
            Err(_) => break, // EOF
        };
        let block_size = u32::consensus_decode(&mut cursor).chain_err(|| "no block size")?;
        let start = cursor.position();
        let end = start + block_size as u64;

        // If Core's WriteBlockToDisk ftell fails, only the magic bytes and size will be written
        // and the block body won't be written to the blk*.dat file.
        // Since the first 4 bytes should contain the block's version, we can skip such blocks
        // by peeking the cursor (and skipping previous `magic` and `block_size`).
        match u32::consensus_decode(&mut cursor) {
            Ok(value) => {
                if magic == value {
                    cursor.set_position(start);
                    continue;
                }
            }
            Err(_) => break, // EOF
        }
        let block: Block = deserialize(&blob[start as usize..end as usize])
            .chain_err(|| format!("failed to parse block at {}..{}", start, end))?;
        blocks.push(block);
        cursor.set_position(end as u64);
    }
    Ok(blocks)
}
```
